### DISCUSSION TITLE
Application architecture using zio

### DISCUSSION THREAD

**Discussion by tusharmath** [zio/zio#4526] [Q&A]
What is the best way to structure a large ZIO application with 100s of components interacting with each other.?

**Comment by jdegoes** [comment]
I would use classic (but "best practice") object-oriented techniques: have components be classes, which depend on other components through constructor parameters, which are interfaces implemented by other components.

The main difference is that for each class, you will make a ZLayer for the constructor for the class, which reflects the dependencies of the component on other components.

This is a modest improvement on the classic object-oriented technique, because it gives you a compile-time, type-safe representation of both the application dependency graph, and the specific way it is wired up in a given configuration.

**Reply by tusharmath** [reply]
Often while building large applications, we end up creating our own functional domains.
How do we structure those domains and interpreters with services/components and what should the dependency tree look like?
More specifically, is onion architecture the way to go?

**Comment by regiskuckaertz** [comment]
@tusharmath Hello - I've been there with my team and one of my devs is a big proponent of the onion architecture, so we decided to give it a try. What we ended up with looks like this:

1. a `logic` SBT project that contains 100% business logic. Any infrastructure-related concern is hidden behind ZIO modules. For all modules, we also define mocks and "laws" checks that any implementation needs to respect.

2. a `infra` SBT project that contains the live implementation for all the modules used in `logic`. This is a departure from the style used in ZIO proper, where you would find a `live` layer in the companion object of the module. We don't do that. We use the cake pattern for each module:

```scala
trait MyModuleLive:
  final def myModule: ZLayer[..., ..., ...] = ???
```

3. finally a `app` SBT project that assembles everything together and provides a main endpoint. In this project you will find objects assembled from all those layers:

```scala
object Live
  extends MyModuleLive
  with MyOtherModuleLive
  ...
```

This has been quite handy as we are using Slick (because reasons 😫) and so our implementations that use database code end up like:

```scala
trait MyDbModule:
  self: Profile with MyTable =>
  import profile.api._
  
  ...
```
and then we can wire the slick profile we want in tests (e.g. `H2Profile`) or in production:

```scala
object Live
  extends MyDbModule
  with MyTable
  with PostgresProfile
```

At first I was a bit reluctant with it, but it grew on me and I quite like that level of separation. The only downside is you will need to define data types that are implementation-agnostic, so there is a bit of repetition there but that's not such a bad thing.
